### ISSUE TITLE
binary_to_float doesn't correctly handle scientific notation with no decimal places specified, such as "1e0"

### ISSUE THREAD

**Issue by Peter-Searby** [erlang/otp#9061]
**Describe the bug**
The binary_to_float built in function should support numbers in scientific notation without decimal places specified.

**To Reproduce**
The following raises a badarg error:
```binary_to_float(<<"1e0">>).```

**Expected behavior**
```1 = binary_to_float(<<"1e0">>).```

**Affected versions**
OTP 27 and all prior versions with this function.

**Additional context**
Other languages, such as C++ and Golang will use this format when all decimal places are 0, so support for this format is important for cross-compatibility.

**Comment by josevalim** [comment]
In my mind, the `binary_to_float` reflects the formats supported by Erlang itself, but I don't know if that's strictly true. If that's the case, would we change the Erlang grammar as well? And this may also require changes to `float_to_binary`:

```
iex(3)> :erlang.float_to_binary(10000000.0, [:short])
"1.0e7"
```

**Comment by richcarl** [comment]
Erlang has held on to the principle "in order to be a floating point number, it has to actually have contain a decimal point", which goes all the way back to its Prolog roots, but you can argue that floating point literals are such a common ground that they should be universally interchangeable across languages. I think it was a design mistake by the C folks to allow the dotless 1e0 version, but since it's there in C, Java, JS, etc., we'd better support it too.

Possibly we could accept the dotless form in the scanner and conversion functions, but let the compiler emit a warning if you actually use that form in code, saying something like "for readability, include the decimal point".

**Comment by richcarl** [comment]
> Currently it lands in `strtod` which should make us compatible with C++, but we have this awkward pre-processing step to allow both `,` and `.` as a decimal separator

Had to test this for myself, because I'd never heard about it, but it's there! But not documented, it seems? Can the commas be dropped?

And if we allow "1e0" we should also allow ".01" (perhaps not warranting a compiler warning).

**Comment by bjorng** [comment]
Yes, I think we should drop support for commas. I didn't know they were allowed.

**Comment by richcarl** [comment]
> There’s a wrinkle to that. If we keep using `strtod` we need to parse the float halfway ourselves anyway so that we can change period into comma on systems whose locale says the decimal point is a comma.

I took a quick look, and only C/C++ strtod cares about locale. Java and Javascript only accept '.'. Seems reasonable to do the same.

**Comment by brigadier** [comment]
> Yes, I think we should drop support for commas. I didn't know they were allowed.

Likely because in quite a few locales comma is the default decimal separator

**Comment by bettio** [comment]
> There’s a wrinkle to that. If we keep using `strtod` we need to parse the float halfway ourselves anyway so that we can change period into comma on systems whose locale says the decimal point is a comma. This is going to be ugly no matter what we do.
> 
> I think we should consider vendoring something like https://github.com/fastfloat/fast_float to get around these issues.

I've found this GitHub issue while trying to fix some problems in AtomVM.

My conclusions here: using `strtod` in our usecase is painful, and vendoring an alternative seems like the only feasible option (also honestly I find all the locale handing that happens inside `strtod` just bad, but that is another story).

Anyway I found an interesting option [here](https://github.com/ulfjack/ryu/tree/master), that looks easy to vendor, and it is under Apache-2.0. They also claim that their implementation is fast and peer reviewed.

There are also some other interesting implementations around that I found: sqlite one (public domain, https://github.com/mackyle/sqlite/blob/ae5d3aa91a794f6f1486b3f453ec44c0aa4c086b/src/util.c#L375-L571) and LuaJIT one (MIT, https://github.com/LuaJIT/LuaJIT/blob/b025b01c5b9d23f6218c7d72b7aafa3f1ab1e08a/src/lj_strscan.c#L15-L327)

**Comment by richcarl** [comment]
> My conclusions here: using `strtod` in our usecase is painful, and vendoring an alternative seems like the only feasible option (also honestly I find all the locale handing that happens inside `strtod` just bad, but that is another story).
> 
> Anyway I found an interesting option [here](https://github.com/ulfjack/ryu/tree/master), that looks easy to vendor, and it is under Apache-2.0. They also claim that their implementation is fast and peer reviewed.

Indeed Ryu is already used in Erlang for the roundtripping printf algorithm: https://www.erlang.org/blog/my-otp-25-highlights/ - if it can completely replace strod then so much better!

**Comment by bettio** [comment]
> > My conclusions here: using `strtod` in our usecase is painful, and vendoring an alternative seems like the only feasible option (also honestly I find all the locale handing that happens inside `strtod` just bad, but that is another story).
> > Anyway I found an interesting option [here](https://github.com/ulfjack/ryu/tree/master), that looks easy to vendor, and it is under Apache-2.0. They also claim that their implementation is fast and peer reviewed.
> 
> Indeed Ryu is already used in Erlang for the roundtripping printf algorithm: https://www.erlang.org/blog/my-otp-25-highlights/ - if it can completely replace strod then so much better!

Actually, after some additional investigation, it looks like Ryu as strtod replacement is experimental (and has some limitations).
I also tried sqlite algorithm, that is easy to port and works on any buffer, but it has a lower precision compared to strtod. 
The right option is using [Eisel-Lemire](https://nigeltao.github.io/blog/2020/eisel-lemire.html) but it needs quite some additional work to port an existing implementation.
